lib/CGameState.cpp

```diff
@@ -381,8 +381,22 @@ void CGameState::CrossoverHeroesList::addHeroToBothLists(CGHeroInstance * hero)
 
 void CGameState::CrossoverHeroesList::removeHeroFromBothLists(CGHeroInstance * hero)
 {
-	heroesFromPreviousScenario -= hero;
-	heroesFromAnyPreviousScenarios -= hero;
+	for (auto obj : heroesFromPreviousScenario)
+	{
+		if (obj->name == hero->name)
```

Comment:
That looks a bit strange to use hero name as an identity. Do we have some hero type or hero identity or whatever? Somehow he knows where to take his icon, base stats name and so on.

Reply:
Also it's interesting, why we have different objects for same hero? Is it normal or some extra object coping somewhere?

Reply:
@ShubusCorporation these are hero objects that comes from json deseralization.
When Campaign scenario ends, the game serializes the heroes into json, and recreate Objects from json in next scenario.
Game takes Heroes from two sources: last map and previous maps. This is why we have 2 lists over here, and some overlapping heroes in both lists. It looks fine, we only need a good way to compare Heroes to say that this is a copy of the same hero. 
For me the hero name is simply "good enough" in our scenarios to find a hero, but no problem, we can look for some systemId instead.

Reply:
So, is `void CGameState::CrossoverHeroesList::addHeroToBothLists(CGHeroInstance * hero)` used anywhere or we should remove it?

Reply:
`subID` field is unique for heroes.

Reply:
there is also `getHeroTypeName()` method, it returns unique key for JSON serialization,